api/services/authService.js

```diff
@@ -0,0 +1,94 @@
+module.exports = {
+  authService: function() {   
+    var https = require("https");
+
+    var clientID = "client_id=" + process.env.CLIENT_ID;
+    var clientSecret = "&client_secret=" + process.env.CLIENT_SECRET;
+    var url = "https://github.com/login/oauth/authorize?";
+
+
+    var redirectToGitHub = function(req, res, state) {
+      res.redirect(url + clientID + "&state=" + state);
+    }
+
+    // Makes post request to github to get accessToken
+    // Calls checkUsers
+    var authenticate = function(req, res, accessCode) {
+      var postOptions = {
+        host: "github.com",
+        path: "/login/oauth/access_token",
+        method: "POST",
+        headers: {
+          "accept": "application/json"
+        }
+      };
+
+      var request = https.request(postOptions, function(response) {
+        response.setEncoding("utf8");
+        var data = "";
+        response.on("data", function(chunk) {
+          data += chunk;
+        });
+        response.on("end", function() {
+          var accessToken = JSON.parse(data).access_token;
+          checkUsers(req, res, accessToken);
+        });
+      });
+
+      request.write(clientID + clientSecret + "&code=" + accessCode);
+      request.end();
+    };
+
+    // Makes get request to github to get user information
+    // Sets req.session.authenticate
+    var checkUsers = function(req, res, token) {
+      var getOptions = {
+        host: "api.github.com",
+        path: "/user?access_token=" + token,
+        method: "GET",
+        headers: {
+          "user-agent": "lucere"
+        }
+      };
+
+      var request2 = https.request(getOptions, function(answer) {
```

Comment:
Pull this functionality out into a utility function. I imagine it's something more generic that you can use. the idea of answer.on("data",...) gives you chunks of data and then you want to git it a callback for what to do on "end"